graylog2-server/src/test/java/org/graylog/testing/mongodb/MongoDBExtension.java

```diff
@@ -0,0 +1,190 @@
+        // in the test class
+        constructInstance(context, Lifecycle.SINGLE_TEST).start();
+    }
+
+    @Override
+    public void afterAll(ExtensionContext context) throws Exception {
+        closeInstance(context);
+    }
+
+    @Override
+    public void afterEach(ExtensionContext context) throws Exception {
+        if (context.getStore(NAMESPACE).get(Lifecycle.class) == Lifecycle.SINGLE_TEST) {
+            closeInstance(context);
+        } else {
+            clearInstance(context);
+        }
+    }
+
+    @Override
+    public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext context) throws ParameterResolutionException {
+        return MongoDBTestService.class.equals(parameterContext.getParameter().getType());
+    }
+
+    @Override
+    public Object resolveParameter(ParameterContext parameterContext, ExtensionContext context) throws ParameterResolutionException {
+        if (MongoDBTestService.class.equals(parameterContext.getParameter().getType())) {
+            return getInstance(context);
+        }
+        throw new ParameterResolutionException("Unsupported parameter type: " + parameterContext.getParameter().getName());
+    }
+
+    @Override
+    public void interceptTestMethod(Invocation<Void> invocation,
+                                    ReflectiveInvocationContext<Method> invocationContext,
+                                    ExtensionContext context) throws Throwable {
+        processFixtures(invocationContext, context);
+
+        invocation.proceed();
+    }
+
+    // Process MongoDBFixtures annotation on methods and classes
+    private void processFixtures(ReflectiveInvocationContext<Method> invocationContext, ExtensionContext context) {
+        ExtensionContext currentContext = context;
+        Optional<MongoDBFixtures> fixtureAnnotation;
+
+        // Find fixture annotation on current element or parent elements
+        do {
+            fixtureAnnotation = findAnnotation(currentContext.getElement(), MongoDBFixtures.class);
+
+            if (!currentContext.getParent().isPresent()) {
+                break;
+            }
+
+            currentContext = currentContext.getParent().get();
+        } while (!fixtureAnnotation.isPresent() && currentContext != context.getRoot());
+
+        if (fixtureAnnotation.isPresent()) {
+            final MongoDBFixtures fixtureFiles = fixtureAnnotation.get();
+            LOG.debug("Loading fixtures {} for {}#{}()",
+                    fixtureFiles.value(),
+                    invocationContext.getTargetClass().getCanonicalName(),
+                    invocationContext.getExecutable().getName());
+            final MongoDBFixtureImporter fixtureImporter = new MongoDBFixtureImporter(fixtureFiles.value(), invocationContext.getTargetClass());
+
+            fixtureImporter.importResources(getInstance(context).mongoConnection().getMongoDatabase());
+        }
+    }
```

Comment:
How about extracting private methods for finding the annotation and loading the fixtures? Would make this block a tad more readable IMO. 

```suggestion
    private void processFixtures(ReflectiveInvocationContext<Method> invocationContext, ExtensionContext context) {
        findFixtureAnnotation(context)
                .ifPresent(annotation -> loadFixtures(invocationContext, context, annotation));
    }

    private Optional<MongoDBFixtures> findFixtureAnnotation(ExtensionContext context) {
        ExtensionContext currentContext = context;
        Optional<MongoDBFixtures> fixtureAnnotation;

        // Find fixture annotation on current element or parent elements
        do {
            fixtureAnnotation = findAnnotation(currentContext.getElement(), MongoDBFixtures.class);

            if (!currentContext.getParent().isPresent()) {
                break;
            }

            currentContext = currentContext.getParent().get();
        } while (!fixtureAnnotation.isPresent() && currentContext != context.getRoot());
        return fixtureAnnotation;
    }

    private void loadFixtures(ReflectiveInvocationContext<Method> invocationContext, ExtensionContext context, MongoDBFixtures fixtureAnnotation) {
        LOG.debug("Loading fixtures {} for {}#{}()",
                fixtureAnnotation.value(),
                invocationContext.getTargetClass().getCanonicalName(),
                invocationContext.getExecutable().getName());
        final MongoDBFixtureImporter fixtureImporter = new MongoDBFixtureImporter(fixtureAnnotation.value(), invocationContext.getTargetClass());

        fixtureImporter.importResources(getInstance(context).mongoConnection().getMongoDatabase());
    }
```